tensorly/decomposition/constrained_parafac.py

```diff
@@ -0,0 +1,284 @@
+        divides all the values by maximum value of the input array.
+    reg_par : float or list of float, optional
+        Amount of regularization (usually called the lambda parameter), applicable for regularized problems only.
+        If there is only one parameter, this parameter will be used for each contraint.
+    prox_par : float or list of float, optional
+        Parameters of the constraints. Refer to each constraint for more info.
+        If there is only one parameter, this parameter will be used for each contraint.
+    cvg_criterion : {'abs_rec_error', 'rec_error'}, optional
+       Stopping criterion if `tol` is not None.
+       If 'rec_error',  algorithm stops at current iteration if ``(previous rec_error - current rec_error) < tol``.
+       If 'abs_rec_error', algorithm terminates when `|previous rec_error - current rec_error| < tol`.
+    fixed_modes : list, default is None
+        A list of modes for which the initial value is not modified.
+        The last mode cannot be fixed due to error computation.
+
+    Returns
+    -------
+    CPTensor : (weight, factors)
+        * weights : 1D array of shape (rank, )
+        * factors : List of factors of the CP decomposition element `i` is of shape ``(tensor.shape[i], rank)``
+    errors : list
+        A list of reconstruction errors at each iteration of the algorithms.
+    References
+    ----------
+    .. [1] T.G.Kolda and B.W.Bader, "Tensor Decompositions and Applications", SIAM
+           REVIEW, vol. 51, n. 3, pp. 455-500, 2009.
+    .. [2] Huang, Kejun, Nicholas D. Sidiropoulos, and Athanasios P. Liavas.
+           "A flexible and efficient algorithmic framework for constrained matrix and tensor factorization." IEEE
+           Transactions on Signal Processing 64.19 (2016): 5052-5065.
+    """
+    rank = validate_cp_rank(tl.shape(tensor), rank=rank)
+    n_modes = tl.ndim(tensor)
+    if constraints is None or len(constraints) != n_modes:
+        constraints = [constraints] * n_modes
+    if reg_par is None or len(reg_par) != n_modes:
+        reg_par = [reg_par] * n_modes
+    if prox_par is None or len(prox_par) != n_modes:
+        prox_par = [prox_par] * n_modes
+
+    weights, factors = initialize_constrained_parafac(tensor, rank, constraints=constraints,
+                                                      reg_par=reg_par,
+                                                      prox_par=prox_par,
+                                                      init=init, svd=svd,
+                                                      random_state=random_state)
+
+    rec_errors = []
+    norm_tensor = tl.norm(tensor, 2)
+
+    if fixed_modes is None:
+        fixed_modes = []
+
+    if tl.ndim(tensor) - 1 in fixed_modes:
+        warnings.warn('You asked for fixing the last mode, which is not supported.\n '
+                      'The last mode will not be fixed. Consider using tl.moveaxis()')
```

Comment:
Yes indeed :) In fact there should always be a non-fixed mode otherwise nothing happens... A simple fix is
- if all modes are fixed, warn user that he is doing something weird and return the init.
- if not all modes are fixed but the last one is, permute a non-fixed mode with the last one at the beginning and at the end.
I do think we need to keep the last mode unfixed during the outer loops since we rely on the last mode update in many algorithms to compute the cost function efficiently.